### DISCUSSION TITLE
-Xss jvm opt for bloop seems to not work

### DISCUSSION THREAD

**Discussion by hilaolu** [scalameta/metals#3350] [Q&A] [answered]
Hi, I am using vscode+metal and I am trying to migrate [a Chisel project](https://github.com/EECS-NTNU/RISCV-FiveStage) from sbt to mill. I got a java stack overflow error during unit test. I have confirmed this is cause by insufficient stack space. The -Xss arg of original sbt project is 32m while the default of bloop is 4m. I have search issues of metal project for how to modify jvm opt. I tried 

- $PROJECT_FOLDER/.bloop/bloop.json
- $PROJECT_FOLDER/.bloop/.jvmopts
- $PROJECT_FOLDER/.jvmopts
- $PROJECT_FOLDER/.mill-jvm-opts
- ~/.bloop/bloop.json
- ~/.bloop/.jvmopts
- Adding -Xss in vscode plugin settings.

And I saw -Xss32m really appeared on java args, with -Xss4m. But the stack overflow problem is still. How can I determind whether the -Xss opt really takes effect?
Thanks

**Comment by dos65** [comment]
@hilaolu setting it in `~./bloop/.jvmopts` should work. 
Could you ensure that after applying modification in this file you restarted the bloop process?

**Reply by hilaolu** [reply]
Restarted many times!

**Comment by kpodsiad** [comment, marked as answer]
EDIT: [Answer to the question is at the end of this thread](https://github.com/scalameta/metals/discussions/3350#discussioncomment-1784425).

The easiest way to achieve your goal is:
- add -Xss to the `~/.bloop/bloop.json`
```json
{
  "javaOptions": ["-Xss32m"]
}
```
or `~/.bloop/.jvmopts`
```
-Xss32m
```
- restart bloop either via metals `restart build server` OR command line (if you installed bloop by yourself) `bloop ng-stop` and then run `connect to the build server` in metals

Then you can search for Bloop process via `ps` or `htop` and see its arguments.

**Reply by kpodsiad** [reply]
@hilaolu We now realized that you didn't ask about bloop options but rather about jvm options passed to unit tests. Unfortunately, there is no such possibility right now :/

**Reply by kpodsiad** [reply]
Do you run those unit tests via code lenses?

**Reply by hilaolu** [reply]
No idea about code lense ( I am new to vscode).
I just click the `test` on class in vscode editor

**Reply by dos65** [reply]
@hilaolu As an option you can try to use [mill-bsp server](https://com-lihaoyi.github.io/mill/mill/Intro_to_Mill.html#_build_server_protocol_bsp)
I don't know if it will work but there is a chance that it runs tests with correct jvm options specified in build.

**Reply by kpodsiad** [reply]
Yeah, so these things are called `code lenses`
<img width="148" alt="Screenshot 2021-12-10 at 10 11 41" src="https://user-images.githubusercontent.com/37124721/145548335-b63060c5-bb81-46d9-af9f-eb9784ec79c4.png">

Currently, as I wrote earlier, there it's not possible to add JVM option to test runned via code lenses if you're using Bloop as build server. However, you could use bloop CLI to run your tests and [add JVM options](https://scalacenter.github.io/bloop/docs/cli/tutorial#pass-jvm-arguments-to-test) to test run. But keep in mind that you have to have Bloop installed ([how to install bloop](https://scalacenter.github.io/bloop/docs/cli/tutorial#pass-jvm-arguments-to-test)).

**Reply by hilaolu** [reply]
Thank you all!

**Comment by kpodsiad** [comment]
@hilaolu It should be possible to define java options in mill like it was done in sbt. Could you try `override def forkArgs = Seq("-Xss32m")`? I checked it myself and it seems to work.

**Reply by hilaolu** [reply]
Hi, @kpodsiad. I didn't modify anything and it works. It seems that `.mill-jvm-opt` or `.jvmopts` works. Thank you and I really enjoy working with metals.
